### DISCUSSION TITLE
Added: LDM Notes for 6/29/2020

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#3627] [LDM Notes]
https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-06-29.md

Agenda
1. Static interface members

I would like to thank @agocke for his continued work on the LDM notes, and hope that I can fill his shoes well as he moves to a new team!

**Comment by YairHalberstadt** [comment]
I suggest a different solution to the `TSelf` problem here: https://github.com/dotnet/csharplang/issues/2528

The key concept is an interface specifies the shape of a type, by giving an example of the type.

Here it would be:

```csharp
interface INumeric TSelf
{
    abstract static TSelf Zero { get; }
    abstract static TSelf operator +(TSelf a, TSelf b);
}
```

One of the advantages of this approach, is it makes HKTs really easy to express:

```csharp
public interface IFunctor TSelf<T>
{
    TSelf<S> Map<S>(Func<T, S> func);
}
```

As well as things like e.g. requiring constructors, nested classes, operators, etc. Without requiring any new syntax.

**Comment by BreyerW** [comment]
For the self problem in most simplistic solution: maybe introduce special keyword `self` that will simply copy type itself? Like

```csharp
interface INumeric<self>
{
     abstract static self Add(...);
}
```

This would under the hood turn into

```csharp
interface INumeric<TSelf> where TSelf : INumeric<TSelf>
{
     abstract static TSelf Add(...);
}
```
Or even more simplistic:

```csharp
interface INumeric<TSelf> where TSelf : self
{
     abstract static TSelf Add(...);
}
```

This alternative way is even self explanatory i think and wouldnt require slighty different codegen when appearing on generics and when on type members

**Comment by yaakov-h** [comment]
That seems to overlap with #902 a bit IMO, though I would love it if that could solve this problem.
